### PR TITLE
eip4844 beacon block proposals

### PR DESCRIPTION
**Pull request by henridf**
Still need to add nim-kzg library and use it to compute the aggregated proof.

### CODE REVIEW

File: beacon_chain/validators/validator_duties.nim [hunk 1]
```diff
@@ -437,6 +440,32 @@ proc getExecutionPayload[T](
...
+  let payload = try:
+    awaitWithTimeout(
+      node.consensusManager.eth1Monitor.getBlobsBundleV1(payload_id),
+      GETBLOBS_TIMEOUT):
+        beacon_block_payload_errors.inc()
+        warn "Getting blobs sidecar from Engine API timed out", payload_id
+        default(BlobsBundleV1)
+  except CatchableError as err:
+    beacon_block_payload_errors.inc()
+    warn "Getting execution payload from Engine API failed",
```
Comment (tersec) [inline beacon_chain/validators/validator_duties.nim]: Mismatched warning message? This is about blobs, not execution payloads

File: beacon_chain/validators/validator_duties.nim [hunk 2]
```diff
@@ -437,6 +440,32 @@ proc getExecutionPayload[T](
       msg = err.msg
     return Opt.some empty_execution_payload
 
+proc getBlobsBundle(
+    node: BeaconNode, epoch: Epoch, validator_index: ValidatorIndex,
+    payload_id: PayloadID): Future[BlobsBundleV1] {.async.} =
+  # https://github.com/ethereum/consensus-specs/blob/v1.1.10/specs/bellatrix/validator.md#executionpayload
```
Comment (tersec) [inline beacon_chain/validators/validator_duties.nim]: Not relevant spec URL?

File: beacon_chain/spec/state_transition.nim [hunk 3]
```diff
@@ -482,10 +486,11 @@ template partialBeaconBlock*(
       voluntary_exits: validator_changes.voluntary_exits,
       sync_aggregate: sync_aggregate,
       execution_payload: execution_payload,
-      bls_to_execution_changes: validator_changes.bls_to_execution_changes
+      bls_to_execution_changes: validator_changes.bls_to_execution_changes,
+      blob_kzg_commitments: kzg_commitments
       ))
 
-proc makeBeaconBlock*[T: bellatrix.ExecutionPayload | capella.ExecutionPayload](
+proc makeBeaconBlock*[T: bellatrix.ExecutionPayload | capella.ExecutionPayload | eip4844.ExecutionPayload](
```
Comment (tersec) [inline beacon_chain/spec/state_transition.nim]: Long line

**Comment by tersec** [comment]
failing on `make wss_sim` -- needs the KZG commitment parameters added there too

**Comment by tersec** [comment]
failing on `make wss_sim` -- needs the KZG commitment parameters added there too

File: beacon_chain/spec/state_transition.nim (hunk 3 above)
Comment (henridf) [reply beacon_chain/spec/state_transition.nim]: fixed

File: beacon_chain/validators/validator_duties.nim (hunk 2 above)
Comment (henridf) [reply beacon_chain/validators/validator_duties.nim]: oops, fixed

**Comment by henridf** [comment]
> needs the KZG commitment parameters added there too

done

**Comment by henridf** [comment]
> Still need to add nim-kzg library and use it to compute the aggregated proof.

My preference would be to do that in a follow-on PR (presumably tomorrow), fwiw. I think in the current state, we'd propose invalid eip4844 blocks, but no behavior should change for pre-eip4844.

**Comment by tersec** [comment]
> > Still need to add nim-kzg library and use it to compute the aggregated proof.
> 
> My preference would be to do that in a follow-on PR (presumably tomorrow), fwiw. I think in the current state, we'd propose invalid eip4844 blocks, but no behavior should change for pre-eip4844.

My concern is that I suspect this won't pass Jenkins testnet finalization at all because of the `minimal` preset issues with EIP4844 you discovered. It hasn't gotten that far yet, and I'm not really sure why. I'm fine with merging this as long as it does something sensible with EIP4844 minimal presets and engine API calls.

(To summarize: `BlobsBundleV1` in https://github.com/ethereum/execution-apis/blob/main/src/engine/experimental/blob-extension.md#blobsbundlev1  assumes a `mainnet` preset blob size, where https://github.com/ethereum/consensus-specs/blob/dev/presets/minimal/eip4844.yaml defines `FIELD_ELEMENTS_PER_BLOB: 4`, which leads to a Blob according to https://github.com/ethereum/consensus-specs/blob/dev/specs/eip4844/polynomial-commitments.md#custom-types having a length of `4 * 32`, not `4096 * 32` which `BlobsBundleV1` requires.)

**Comment by tersec** [comment]
I agree, `nim-kzg` should be in another PR though, not this one.

**Comment by henridf** [comment]
> something sensible with EIP4844 minimal presets and engine API calls.

Following your suggestion, made a change 2420d69011879610ae012b4a078091bc7fe99c5f to skip blob fetching when the preset is minimal. This means that all block proposals will be invalid for EIP4844-minimal, but as we've agreed elsewhere, EIP4844 is mainnet-preset only for now.

**Comment by tersec** [comment]
> > something sensible with EIP4844 minimal presets and engine API calls.
> 
> Following your suggestion, made a change [2420d69](https://github.com/status-im/nimbus-eth2/commit/2420d69011879610ae012b4a078091bc7fe99c5f) to skip blob fetching when the preset is minimal. This means that all block proposals will be invalid for EIP4844-minimal, but as we've agreed elsewhere, EIP4844 is mainnet-preset only for now.

Seems fine, yeah.

Will wait to see CI a bit, but inclined to merge at this point, and can build on it in future PRs.

File: scripts/launch_local_testnet.sh [hunk 4]
```diff
@@ -785,6 +785,8 @@ DEPOSIT_CONTRACT_ADDRESS: ${DEPOSIT_CONTRACT_ADDRESS}
 ETH1_FOLLOW_DISTANCE: 1
 ALTAIR_FORK_EPOCH: 1
 BELLATRIX_FORK_EPOCH: 2
+CAPELLA_FORK_EPOCH: 3
```
Comment (tersec) [inline scripts/launch_local_testnet.sh]: This works, Even without separate VC support?

Comment (tersec) [reply scripts/launch_local_testnet.sh]: I did this and it worked fine a while ago but only after disabling the separate VC validators and only using the built-in BN validators.

Comment (henridf) [reply scripts/launch_local_testnet.sh]: Oops, this is unintentional. Juggling between branches and my own tests. Thanks for the catch.

Comment (henridf) [reply scripts/launch_local_testnet.sh]: 311ddd19fd0de3fd57837e5d2b23d60f6554c2de

**Review by tersec — APPROVED** [review]

**Pull request merged**